### PR TITLE
Fixes issue with CSV detection with sufficiently sparse columns & rows

### PR DESCRIPTION
**Pull request by lettergram**
This PR fixes an incorrect counting mechanism where by two delimiter counts were subtracted from the total discovered. Particularly, to account for the first and last finding in the line (per the comment). Once this offending code was removed all tests pass; previous to code removal the `sparse-columns-test.csv` was identified as a text file. 

- Added 3 new test files
    - `sparse-first-column.txt` - tests sparsely populated first columns.
    - `sparse-first-and-last-column.txt` - tests sparsely populated first & last column.
    - `sparse-columns-test.csv` - tests sparsely populated columns, particularly where rows for some columns stop near the top of the file.
    - `sentence-4x.txt` - tests if four sentences ending in a period are labeled as a csv (more on why this test was added below).
    - `names-col-empty.txt` tests if theres a large number of empty rows for a single column file, will the deliminator be properly identified (skipping empty rows).

Due to this fix, it's possible to increase the threshold on the `max_count_percent`. The only file that gets close to triggering the 0.85 threshold is `names-col.txt` this is due to lines 214-215 in `data_profiler/data_readers/csv_data.py`:

```python
# If no delimiter, see if spaces are regular intervals                    
count = line.count(" ")
```

6/7 of the lines in the file have spaces, only the header does not:

```
NAMES
George Washington
John Adams
Thomas Jefferson
James Madison
James Monroe
John Adams
```

However, if you reduced the names to four, the function would fail to identify the file as a CSV.

```
NAMES
George Washington
John Adams
Thomas Jefferson
James Madison
```

Further, blanks spaces were also being "counted" as zero delimiters. While I think that's fair, it's also a "blank row" in the case of a single element column. Thus, I added a check to ensure greater than one (`\n`) character was present and a test (which would fail pre-change):

```
NAMES
George Washington




John Adams
Thomas Jefferson
James Madison








James Monroe
```

Finally, and perhaps most controversially, I modified two of the "text" tests (`sentence-3x.txt` and `sentence-10x.txt`). Essentially, the CSV sniffer was picking up `.` as a separator. There were two choices, remove it as a separator OR have a string of sentences ending in a `.` to be labeled as a CSV. 

Personally, I think a series of sentences should be labeled as a CSV (looks that way to me). I know this was discussed in the past, but given the current bugs, it warrants another review. 

I thus modified the tests slightly (just using different punctuation) to have the tests pass. While this may be controversial it does match what the CSV reader library already does. For consistency, I added the prior test to the CSV test suite. This ensures we are explicit about how the library is handling this case.

### CODE REVIEW

File: data_profiler/data_readers/csv_data.py [hunk 1]
```diff
@@ -236,8 +237,8 @@ def is_match(cls, file_path, options=None):
         max_count_percent = count_percent[max_count_index]
         
         # Infered the file was a CSV
-        if ((max_count_value > 0 or delimiter is None) \
-            and max_count_percent >= 0.8):
+        if (max_count_value > 0 or delimiter is None) \
+            and (max_count_percent >= 0.80):
```
Comment (JGSweets) [inline data_profiler/data_readers/csv_data.py:241]: Would you indent this one over? Will make it easier to read.

File: README.md [hunk 2]
```diff
@@ -236,6 +236,11 @@ To run file of unit tests, use form:
 python3 -m unittest discover -p test_profile_builder.py
 ```
 
+To run a file with Pytest use:
+```
+pytest data_profiler/tests/data_readers/test_csv_data.py -v
```
Comment (AnhTruong) [inline README.md:241]: why do we need pytest here? if so, we have to make it for other test use cases, e.g., all tests

File: data_profiler/data_readers/csv_data.py [hunk 3]
```diff
@@ -196,38 +196,39 @@ def is_match(cls, file_path, options=None):
...
                 delimiter_locs = [i.start() for i in re.finditer(delimiter_regex, line)]
                 if delimiter:
                     # Count all locations that aren't the last entry
                     # This can cause errors, if you could have a sparse end column
-                    count = 0
                     for loc in delimiter_locs:
-                        # Reason: len(line) - 2; -1 to start count 0 & -1 for last loc
-                        if loc != len(line)-2:
-                            count += 1
+                        count += 1
```
Comment (AnhTruong) [inline data_profiler/data_readers/csv_data.py:211]: for now, why don't just do len(delimiter_locs) ?

File: data_profiler/tests/data/csv/names-col-empty.txt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+
+
+
+
+
+
+
+James Monroe
+John Adams
```
Comment (AnhTruong) [inline data_profiler/tests/data/csv/names-col-empty.txt:19]: do we have files with blank lines at the last lines?

**Review by AnhTruong — COMMENTED** [review]
some comments

File: README.md (hunk 2 above)
Comment (lettergram) [reply README.md:241]: Just an example of an alternative method. Pytest is part of the requirements-test.txt

File: data_profiler/tests/data/csv/names-col-empty.txt (hunk 4 above)
Comment (lettergram) [reply data_profiler/tests/data/csv/names-col-empty.txt:19]: That's actually how I discovered this issue (big file of newline characters, nothing else).

I can add it to the end of this file.

File: data_profiler/data_readers/csv_data.py (hunk 3 above)
Comment (lettergram) [reply data_profiler/data_readers/csv_data.py:211]: Fair, PR'd update

File: data_profiler/data_readers/csv_data.py (hunk 1 above)
Comment (lettergram) [reply data_profiler/data_readers/csv_data.py]: Done, please re-review

File: data_profiler/data_readers/csv_data.py [hunk 5]
```diff
@@ -188,56 +188,72 @@ def is_match(cls, file_path, options=None):
...
                 else:
                     # If no delimiter, see if spaces are regular intervals
                     count = line.count(" ")
-                    
-                if count in delimiter_count:
-                    delimiter_count[count] = delimiter_count[count] + 1
-                else:
-                    delimiter_count[count] = 1
+
+                # Track the dilimiter count per file
```
Comment (JGSweets) [inline data_profiler/data_readers/csv_data.py:219]: spelling

**Review by AnhTruong — APPROVED** [review]

File: data_profiler/data_readers/csv_data.py [hunk 6]
```diff
@@ -188,56 +188,72 @@ def is_match(cls, file_path, options=None):
                 options.update(header=cls._determine_has_header(data_as_str))
         
         max_line_count = 1000
-        min_line_count = 2
+        min_line_count = 3
```
Comment (lettergram) [inline data_profiler/data_readers/csv_data.py:191]: Now you need at least 3 lines to match a CSV.

**Comment by lettergram** [comment]
Added `incorrect-csv.txt` as a text data_reader check, to ensure the CSV `is_match` function doesn't trigger on it.

File: data_profiler/tests/data/txt/incorrect-csv.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+a b c
+"d e f"
+h i j
+"k l m"
+"n o p"
+q r s
+t u v
+w x y
+z 1 2
```
Comment (lettergram) [inline data_profiler/tests/data/txt/incorrect-csv.txt:9]: By default CSV reader will read it as:

```python
>>> with open('data_profiler/tests/data/txt/incorrect-csv.txt') as csvfile:
...     r = csv.reader(csvfile)
...     for row in r:
...             print(', '.join(row))
... 
a b c
d e f
h i j
k l m
n o p
q r s
t u v
w x y
z 1 2
```
```python
>>> with open('data_profiler/tests/data/txt/incorrect-csv.txt') as csvfile:
...     r = csv.reader(csvfile)
...     for row in r:
...             print(row)
... 
['a b c']
['d e f']
['h i j']
['k l m']
['n o p']
['q r s']
['t u v']
['w x y']
['z 1 2']
```

**Review by gme5078 — APPROVED** [review]

**Review by AnhTruong — DISMISSED** [review]

**Review by gme5078 — APPROVED** [review]

**Review by gme5078 — APPROVED** [review]

**Pull request merged**